### ISSUE TITLE
git pile format-patch cover letters are unreliable

### ISSUE THREAD

**Issue by ashutoshx** [git-pile/git-pile#41]
Every once is a while (though this is fairly rare) we hit a situation where a 'git pile am -g' with a cover letter generated with 'git pile format-patch' results in a branch which is different from the original branch used to generate the cover letter.

As we know 'git pile format-patch' cover letters are generated using 'git range-diff' which has its own algorithm for determining that patches are identical and actually takes a --creation-factor argument for the algorithm.

One instance we have seen this happen is when a patch was identical to what it was previously except for a change in line numbers due to large amount of code deletion in the file (which was part of a different patch). Therefore 'git range-diff' considered this patch as unchanged and consequenty the patch was applied in the wrong place in the file during 'git am' during 'git pile genbranch'.

So that is one example where a 'git range-diff' based git pile cover letter yields erroneous results. I am not sure if there are multiple such instances where we see this issue due to the use of git range-diff (e.g. if the diff produced during 'git range-diff' has similar issues or is that always the same as the output of a regular diff).

Apart from different final and original branches I believe in some rare cases cover letters have also led to git pile genbranch failures because the patches were updated incorrectly and therefore could not be applied.

I am creating this issue to see if anything can be done about this. Ideas include:

* Use git range-diff to only include the "metadata" in the cover letter, but include the actual diff of the patches which allows the patches to always be reconstructed accurately. Whether this causes other problems like git am conflicts or very large cover letters with say a lot of line number diff's in the patches remains to be seen.

* Actually iterate: Follow the 'git pile format-patch' with 'git pile am -g' with the generated cover letter and make sure the output branch is the same as the original branch we started with and continue the process till this happens.

Also not sure if passing --creation-factor argument to 'git pile format-patch' helps in such situations and this is enough to address such situations.

**Comment by lucasdemarchi** [comment]
Actually I don't think the real problem is with the filtering from git range-diff. We only use range-diff to **filter** the diff that we send in the cover-letter. So if the final result is wrong and the fault is git-range-diff's fault, then you'd see the final output missing parts of the other branch, instead of seeing it applied to the wrong location.

One issue that we have is that the branch interdiff doesn't match exactly what you have in your patches/ directory, because after several patches applied the hunk numbers start to drift. What I've seen is that in this case, with very similar functions, `git am` may end up applying the patch to the wrong place.  So, I wouldn't say it's git-range-diff's fault or that the creation-factor would change anything: the problem is actually the stale data we have saved in the patches directory.

Ideally we would have https://github.com/git-pile/git-pile/issues/10, allowing to patch a branch without resorting to the physical patch files, but AFAIK there isn't anyone working on this in upstream git. I've seen some talks about having it, but it was a long time ago.

For what we currently have with the physical patches, it's a good idea to "refresh" them with the current state of the branch from time to time. However, forcing format-patch to always send the complete diff would increase the noise a lot and create a lot of conflicts. I think in the very beginning

**Comment by ashutoshx** [comment]
> Actually I don't think the real problem is with the filtering from git range-diff. We only use range-diff to filter the diff that we send in the cover-letter. So if the final result is wrong and the fault is git-range-diff's fault, then you'd see the final output missing parts of the other branch, instead of seeing it applied to the wrong location.

I can provide an example of how this happens (I have already provded with some internal code which is not public). Also, as I mention above, another thing we have noticed is that with the cover letter the patches are updated incorrectly and therefore 'git am' fails to apply them so genbranch fails.

> One issue that we have is that the branch interdiff doesn't match exactly what you have in your patches/ directory, because after several patches applied the hunk numbers start to drift. What I've seen is that in this case, with very similar functions, git am may end up applying the patch to the wrong place. So, I wouldn't say it's git-range-diff's fault or that the creation-factor would change anything: the problem is actually the stale data we have saved in the patches directory.

The problem with filtering with 'git range-diff' is that the diff in the cover letter is no longer an exact diff (or a "lossless" diff), it is actually a lossy diff meaning it's losing information (as evidenced by the --creation-factor argument to 'git range-diff')). The drift in the hunk numbers is itself a result of this, because 'git range-diff' is considering patches as unchanged even when hunk numbers in the patches are changing, hunk line numbers drift over time.

> For what we currently have with the physical patches, it's a good idea to "refresh" them with the current state of the branch from time to time. However, forcing format-patch to always send the complete diff would increase the noise a lot and create a lot of conflicts.

Has this ever been tried? The diff to include in the cover letter should be not be filtered with 'git range-diff' but should be the exact diff produced by 'git pile genpatches'. Conflicts may increase but it just means the cover letter is stale and must be updated prior to merging. Though the drift in the hunk numbers will stop so not sure if that will help. But I agree this has to be tried out and see what the trade-off here would be, though even then I'd think it is hard to argue against correctness.

In any case we can continue with the current cover letters but I wanted to file this issue as a reference to point people to when they see these issues due to the use of lossy 'git range-diff' filtering.

**Comment by lucasdemarchi** [comment]
> The problem with filtering with 'git range-diff' is that the diff in the cover letter is no longer an exact diff (or a "lossless" diff), it is actually a lossy diff meaning it's losing information (as evidenced by the --creation-factor argument to 'git range-diff')). The drift in the hunk numbers is itself a result of this, because 'git range-diff' is considering patches as unchanged even when hunk numbers in the patches are changing, hunk line numbers drift over time.
> 
> >

sorry, but creation-factor isn't evidence of anything: it only tells git if something should be considered a change or should be considered something new. I.e. if git should interpret that when creating the diff as an add + remove or a change.

From what I can see, the diff produced with the filtering would be an exact diff if the hunk numbers stored in the patches hadn't drifted. I think we will need a synthetic example here, otherwise we will just have different opinions on what the issue is.

> Has this ever been tried? The diff to include in the cover letter should be not be filtered with 'git range-diff' but should be the exact diff produced by 'git pile genpatches'. Conflicts may increase but it just means the cover letter is stale and must be updated prior to merging. Though the drift in the hunk numbers will stop so not sure if that will help. But I agree this has to be tried out and see what the trade-off here would be, though even then I'd think it is hard to argue against correctness.

internally, yes. We survived I think 2 weeks without the filter. Things we can do if we want to try it again:
1) Add a a flag in format-patch to trigger this behavior.
2) Add code to a bot to do a "patch refresh" following each push. This should be easy with github actions now that #24 is ready for prime-time. This will tell us if we start to have conflicts due to the hunks diff
3) I actually see #10 as a more longterm solution, so we stop relying on the physical patches and we instead have a proper format in git itself to translate one branch to another.

> In any case we can continue with the current cover letters but I wanted to file this issue as a reference to point people to when they see these issues due to the use of lossy 'git range-diff' filtering.

I'm yet to see evidence to blame git-range-diff here. All the examples I saw in the past (which includes looking to our internal failure examples) were not because git range-diff produced something wrong. I will take a look in the one you sent me last week, but I have high confidence it falls in the same bucket: if fails because the base in which it was applied had already drifted (yes, as a result of previously . Remember that git-range-diff is only used to **remove** parts of the overall diff. So if something is applying to the wrong hunk, it's not git-range-diff's fault. It would be if something was present in the result-branch, but not included in format-patch.

**Comment by lucasdemarchi** [comment]
I was able to reproduce it on https://github.com/git-pile/playground. As I said, it doesn't seem to be related with any filtering by git range-diff. I hand crafted 2 changes to force a silent conflict in a file due to the hunk line numbers. See the below branches:
- https://github.com/git-pile/playground/tree/feature1-d8f930aea4c1
- https://github.com/git-pile/playground/tree/feature2-d8f930aea4c1

d8f930aea4c1 is the pile commit on top of which they were developed. If you apply feature1 and then feature2 you will have the wrong result. The second change will end up applied in the wrong block. That is because feature1 moved the entire block up, and both blocks had exactly the same number of context lines.  See the wrong result at https://github.com/git-pile/playground/blob/internal-next-feat-1-2/test-hunk-numbers-fail.txt

I also pushed the result for when the "features" are applied in the opposite order: https://github.com/git-pile/playground/blob/internal-next-feat-2-1/test-hunk-numbers-fail.txt

And the diff between them:

```diff
$ git diff origin/internal-next-feat-2-1..origin/internal-next-feat-1-2
diff --git a/test-hunk-numbers-fail.txt b/test-hunk-numbers-fail.txt
index 65edab4..ee5b530 100644
--- a/test-hunk-numbers-fail.txt
+++ b/test-hunk-numbers-fail.txt
@@ -16,7 +16,7 @@ block 2 {
        This is a paragraph that repeats itself multiple times. 2
        This is a paragraph that repeats itself multiple times. 3
        This is a paragraph that repeats itself multiple times. 4
-       This is a paragraph that repeats itself multiple times. 5a
+       This is a paragraph that repeats itself multiple times. 5
        This is a paragraph that repeats itself multiple times. 6
        This is a paragraph that repeats itself multiple times. 7
        This is a paragraph that repeats itself multiple times. 8
@@ -29,7 +29,7 @@ block 3 {
        This is a paragraph that repeats itself multiple times. 2
        This is a paragraph that repeats itself multiple times. 3
        This is a paragraph that repeats itself multiple times. 4
-       This is a paragraph that repeats itself multiple times. 5
+       This is a paragraph that repeats itself multiple times. 5a
        This is a paragraph that repeats itself multiple times. 6
        This is a paragraph that repeats itself multiple times. 7
        This is a paragraph that repeats itself multiple times. 8
```

**Comment by lucasdemarchi** [comment]
Also, looking at this problem closely, it doesn't seem related to the use of interdiff: same issue happen just inverting the order of patches in a normal non-pile branch.

**Comment by ashutoshx** [comment]
Sorry about the creation factor, you are right, I was confusing that with a "cost factor" which I think is in play for git range-diff to determine that a patch hasn't changed.

Thanks also for the examples, I agree with what you are saying and also the analysis below in the email:

         - we drop the object hash from the individual diff hunks due to
          the conflicts this could cause. In other words, we have a
          plain diff rather than a git diff stored in the patches/
          directory

        - Due to not having the object hashes, git never does the
          fallback to 3-way patch application and it instead relies on
          the line numbers and the context in the patch

        - If you have 2 different places sufficiently similar and line
          numbers that drifted over time, git will apply the patch to
          the wrong location

But I still think the root cause is still the filtering with git range-diff. In the example we are considering there is a single patch in which one hunk is applied at the wrong place. That patch is considered as unmodified (presumably due to filtering with git range-diff) and not included the cover letter. Now if that patch was considered modified and the change in the hunk line numbers were included in the cover letter the patch would have been applied correctly.

However we know that we cannot start including these hunk number changes in the cover letter (that is stop filtering with git randge-diff for patches which are otherwise identical) because if we do that we will run into the other problem of causing many conflicts when people apply stale cover letters (a use case which we want to support). So I guess let us continue with what we are doing since these kinds of issues are relatively rare and getting rid of git range-diff filtering will cause greater problems.

Maybe we should close this issue as "will not fix".

**Comment by lucasdemarchi** [comment]
I think there is still value in the follow semi-related task:
1) Teach format-patch to work without filters. This should allow people to share a complete patch if they need. It would basically be the equivalent of `genpatches; cd patches; git commit -a; git format-patch -1`, but a little more automated.
2) Teach genpatches to apply filter (or should we use something like --stdout to format-patch?) so it's easier to experiment and apply local changes

For long-term what I think would fix it would be #10, but that is far from happening.
